nn/transformer.py

```diff
@@ -0,0 +1,237 @@
+"""
+Transformer Modules
+"""
+
+from __future__ import annotations
+from typing import Optional, Any, Union, Callable
+from .. import nn
+
+
+class TransformerEncoderLayer(nn.Module):
+  """
+  Defines one layer of a standard transformer encoder
+  """
+  def __init__(self, dim_model: int, num_heads: int, dim_ff: int = 2048, dropout: float = 0.1,
+               activation: Callable[[nn.LayerRef], nn.LayerRef] = nn.relu, layer_norm_eps: float = 1e-5,
+               norm_first: bool = False) -> None:
+    """
+    :param dim_model: hidden dim, PyTorch name: d_model
+    :param num_heads: number heads, PyTorch name: nhead
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param dropout: Dropout value, PyTorch name: dropout
+    :param activation: activation function
+    :param layer_norm_eps: Epsilon value for layer normalization
+    :param norm_first: if ``True`` will perform normalization before other att and ff operations, otherwise after
+    """
+    super().__init__()
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # to be implemented
+
+    self.linear_ff = nn.Linear(dim_ff)
+    self.linear_out = nn.Linear(dim_model)
+
+    self.activation = activation
+    self.norm_first = norm_first
+    self.norm_eps = layer_norm_eps
```

Comment:
Why is this not consistent?

Reply:
`layer_norm_eps` vs `norm_esp`

Reply:
yes you are right, changing to layer norm, since we always want to do (some) layer norm here.

Reply:
But see my other comments. Why do you have `norm` as an option in other places? This is also inconsistent. Either have it configurable everywhere in the same way, or nowhere.
Also, in any case, I would probably prefer `norm_eps`.

Reply:
Ahh you mean why it is an option in TransformerEncoder/Decoder but not in the layer itself?

Reply:
There is no reason, I can add that. Also If you prefer `norm` instead of `layer_norm` then we cut the `layer_prefix` from the variables.